### DISCUSSION TITLE
`notifyOnChangeProps` to also accept a callback

### DISCUSSION THREAD

**Discussion by paulobmarcos** [TanStack/query#5588] [Ideas]
Hey TanStack team! 

First of all, thank you for your amazing work with the TanStack/query 🚀

## Problem

With `notifyOnChangeProps` we can make a component only re-render if any of the listed properties change. This is very useful and we can completely stop all re-renders if needed with an empty array `[]`.

We have a use case where we want to completely disable re-renders that would come from data fetching when components go into an "hidden"/"sleep" state.

This is achievable by setting component state and use that as a condition to pass in a different `notifyOnChangeProps` value.

Basic Example:
```tsx
const useSharedDataHook = (options) =>
  useQuery({
    queryKey: ['myExampleKey'],
    queryFn: async () => {
      const response = await fetch(
        'https://api.github.com/repos/tannerlinsley/react-query'
      );
      return response.json();
    },
    ...options,
  });

const MyExampleComponent = () => {
  const [hidden, setHidden] = useState(false);
  const { dataUpdatedAt } = useSharedDataHook({
    notifyOnChangeProps: hidden ? [] : undefined,
  });

  const toggleHidden = () => setHidden((h) => !h);

  return (
    <>
      <div>DataUpdatedAt: {dataUpdatedAt}</div>
      <button onClick={toggleHidden}>Toggle Hidden</button>
    </>
  );
};
```

So, if any other component uses the same `queryKey`  to update the data, the component won't re-render.

However, this still requires one extra re-render to change the `notifyOnChangeProps` value and we have a use case where the extra re-render could be considered expensive.

## Proposal

The `notifyOnChangeProps` would be extended to also accept a callback. The return type of this callback would be `string[] | "all"`.

This way, we could change the return value of `notifyOnChangeProps` without an extra re-render.

Basic Example:
```tsx
const MyExampleComponent = () => {
  const hiddenRef = useRef(false);

  const notifyOnChangeProps = useCallback(() => {
    return hiddenRef.current ? [] : undefined;
  }, []);

  const { dataUpdatedAt } = useSharedDataHook({ notifyOnChangeProps });

  const toggleHidden = () => {
    hiddenRef.current = !hiddenRef.current;
  };

  return (
    <>
      <div>DataUpdatedAt: {dataUpdatedAt}</div>
      <button onClick={toggleHidden}>Toggle Hidden</button>
    </>
  );
};
```

## Unknowns

I'm still unsure if there would be any side-effects by opting for a callback instead.

From the example above, the only probable side-effect I'm able to think of is that the data wouldn't be immediately up to date if the component stopped being hidden, probably requiring a manual "refetch" or wait for another data fetch.

**Comment by TkDodo** [comment]
I think the react way would be to unmount the component if it's going to be "hidden"

**Reply by paulobmarcos** [reply]
Thanks for the reply @TkDodo!

In some circumstances unmounting can be a solution, but I believe it depends on the use case.

A use case where this would be useful, and unmounting not a "good" solution, is with React-Native and non-active stacked screens:
![image](https://github.com/TanStack/query/assets/11398440/3ff1f190-801f-4620-9849-4f5be13e99d2)

Non-active screens still re-render even though they are not visible.

The proposal would help this non-active screens to not re-render on data updates, caused by active ones, without requiring an extra re-render to set an "active"/"inactive" state.

This idea would extend to any other scenario that would benefit from "pausing" re-renders regarding data updates.

**Reply by TkDodo** [reply]
so is it a performance concern or why can't the components just render? Also, why not unmount just the children:

```
function Component() {
  const { data } = useQuery(...)

  return hidden ? null : <div>...</div>
}
```

**Reply by ferrannp** [reply]
@TkDodo the problem here is more React Native than React. In React Native we have the concept of screens behind other screens (navigation). React Query firing in screens that are not visible is in a lot of cases not desirable and bad for React Native performance. We could use this hook https://reactnavigation.org/docs/use-is-focused/ (which is terribly bad for performance too) in this way:

```js
 const isFocused = useIsFocused();
 const { dataUpdatedAt } = useSharedDataHook({
  notifyOnChangeProps: isFocused ? [] : undefined,
 });
```

The problem is that this hook gets triggered when you blur a screen (and focus), causing a re-render that is very heavy on React Native (nothing should happen on navigation). That is why there is a warning in the docs:

> Note that using this hook triggers a re-render for the screen when it changes focus. This might cause lags during the animation if your screen is heavy

The proposed solution is:

https://reactnavigation.org/docs/use-focus-effect/

Now, if we cannot use a `useFocusEffect` to enable/disable subscriptions to `react-query` using `useQuery`... Then, with the callback at least we could do:

```js
const notifyOnChangeProps = useCallback(() => {
  return navigation.isFocused() ? [] : undefined;
}, [navigation]);

const { dataUpdatedAt } = useSharedDataHook({ notifyOnChangeProps });
```

Not only `notifyOnChangeProps`, but `enabled` could also accept a callback.

Edit:

```js
function Component() {
  const { data } = useQuery(...)

  return hidden ? null : <div>...</div>
}
```

^ this would also hit performance and look bad, as you can swipe back to see the content behind. We should really not alter the screen behind when it is not focused (or during navigation).

**Reply by paulobmarcos** [reply]
Thanks @TkDodo and @ferrannp for your inputs.

Yes, animating a component in and out can be a good example when we still have to render the components children :+1:

**Reply by TkDodo** [reply]
Thanks for outlining the react-native world for me, that's really interesting.

I think we can make `notifyOnChangeProps` a function, but I wouldn't let it return `undefined` (which is for tracking props). This would mean changes would be confined here:

https://github.com/TanStack/query/blob/457723f2cd47a460d351d695d94bff962e7f68e5/packages/query-core/src/queryObserver.ts#L591-L600

Would you like to contribute it? What we'd need is:

- the change
- a test in the `query-core` package
- updated API docs
- updated react-native docs to show this pattern, as it will be increasingly useful for those users.

What do you think?

**Reply by paulobmarcos** [reply]
Thanks @TkDodo! Yes, I agree, the callback should return `string[] | "all"`.

I'm willing to contribute to make this happen 👍

I have just a question though. Since this doesn't cause a re-render, when we change the `ref`, we might have to trigger an update manually. Probably calling `refetch`?

**Reply by TkDodo** [reply]
the `notifyOnChangeProps` function would only run after a component has fetched data. Note that if your component renders top-down, because of a state change in a parent, `notifyOnChangeProps` cant't "opt out" of rendering. This is only to avoid renders that would be triggered by react query itself.

**Reply by paulobmarcos** [reply]
@TkDodo yes, of course, this would just stop re-renders triggered by the `useQuery` hook 👍 

What I was trying to say is that with a state update, we are able to get the latest changes from cache when we revert `notifyOnChangeProps` from `[]` to `all`.

Example video:
https://github.com/TanStack/query/assets/11398440/6f3e0e97-a95e-4c7c-83c9-93ad3d2fda2e

With the callback example, we might need to manually "refetch" to get the latest data when a component is focused again, but I believe that should be fine.

**Reply by TkDodo** [reply]
If you go from one screen to the next, and your component re-renders, it will pick up the new `notifyOnChangeProps` value. If your component doesn't re-render, then yeah, it would only pick it up with a new refetch. is that where `useFocusEffect` comes in now:
- https://tanstack.com/query/v4/docs/react/react-native#refresh-on-screen-focus

I'm still a bit confused to what is the desired behaviour. It seems like the version with `useState` from the video works fine, too?

**Reply by ferrannp** [reply]
I agree here that it would be the app's responsibility to refetch on `useFocusEffect` when necessary.

The desired behavior here is that we want to be able to stop `useQuery` to being triggered when we don't want to and, in order to do that we want some kind of unsubscribe pattern (or the function method described here) instead of re-rendering the screen in order to "unsubscribe". Hopefully I'm explaining myself correctly ☺️

**Reply by TkDodo** [reply]
ah okay got it. Sooo, there is another thing that actually stops subscriptions:

https://github.com/TanStack/query/blob/457723f2cd47a460d351d695d94bff962e7f68e5/packages/react-query/src/useBaseQuery.ts#L62-L72

If we are in `restoring` state, components render, but are not subscribed to the QueryCache. When can we be `restoring` ? Well, we have our own Provider for that that has a boolean value:

https://github.com/TanStack/query/blob/ba7fc56440ddd24a756be25894fe87134c0fed22/packages/react-query/src/isRestoring.ts#L4-L7

We currently only use this context to signal that we are restoring from our persist plugins:

https://github.com/TanStack/query/blob/a6186e50488bd3e14b676bf4aedd0953be3f0574/packages/react-query-persist-client/src/PersistQueryClientProvider.tsx#L54

The Provider and the hook are exported, so what you could try is:

- render this Provider on your screen
- provide it `true` when you want to unsubscribe
- set it back to `false` when you want to subscribe

It's not what we have intended the provider for, at least according to the name, but if unsubscribing from queries while keeping data rendered on the screen is a real use-case, I'd rather push this as a solution rather than the `notifyOnChangeProps`, given that it works for you.

If that's the case, I could consider a re-name in v5 (because you are not restoring anything here).

**Reply by ferrannp** [reply]
Mmm I see @TkDodo, the problem with that though (again) is that we need to trigger a re-render of the screen (or the wrapper that wraps the screen) to change the value for the provider, correct? And that will happen when navigating in RN, and anything that happens while navigating will drop frames. Is there a way we could do something like this:

```js
const {subscribeToQuery, unsubscribeToQuery} = useQuery(...)

useFocusEffect(useCallback(() => {
  subscribeToQuery();
  return () => {
     unsubscribeToQuery();
  };
},[subscribeToQuery, unsubscribeToQuery])
```

While still using `useQuery` and without forcing any extra re-renders?

This would give us tremendous amount of control over the subscriptions without causing performance issues in RN.

**Reply by TkDodo** [reply]
okay, back to `notifyOnChangeProps` it is :)

>  Is there a way we could do something like this:

This is pretty much what `useQuery` does under the hood. You could create your own custom hook that also creates a `QueryObserver` and then subscribes / unsubscribes in `useFocusEffect` instead of `useSyncExternalStore`. `QueryObserver` is the primitive from query-core that has `subscribe / unsubscribe` methods

**Reply by ferrannp** [reply]
Yup. We were thinking though how to contribute so we do not have to maintain our own hooks. I was thinking if you would be ok to have a `useQuery.native.ts`so, basically the `react-native` packager would pick that (and web will use the normal `useQuery.ts`). Then, we could have some arg like `unsubscribeOnBlur` (maybe even others like `unsubscribeOnBackground` - we can thing of all those) and for example, if you pass  `unsubscribeOnBlur` and you do not have `react-navigation` installed, you would see a warning (in web, that prop would be ignored). But I was wondering if changing `useQuery` like creating a `useQuery.native` would be enough or if we would need to touch other hooks too?

**Reply by TkDodo** [reply]
so this would mean we'd need a copy of every hook with a `.native.ts` extensions, where we go away from subscribing in `useSyncExternalStore` and switch towards `useFocusEffect`. I'm not sure I like that too much from a) the maintenance effort and b) react concurrent features really need us to `useSyncExternalStore` or we risk tearing.

I think the `notifyOnChangeProps` function is still the best move

**Comment by paulobmarcos** [comment]
Hey @TkDodo! I have a branch ready and would like to open a PR.

However I might not have permissions to push to a branch. What is the best away to do so? 🙏

**Reply by TkDodo** [reply]
create a fork, create a branch there and then create a PR from your fork to our repo :)

**Comment by paulobmarcos** [comment]
Thank you so much @TkDodo for merging the PR: https://github.com/TanStack/query/pull/5734

---

A few more questions that could further improve this experience.

We tried this approach, since it's now live, and we hit what I described here: https://github.com/TanStack/query/discussions/5588#discussioncomment-6233477

Whenever a fetch happens, if a component is not listening to any prop changes it could end up with "outdated" data.  
Here is an example on React Native:
https://github.com/TanStack/query/assets/11398440/61d80168-ef5f-4973-911f-848ab1eb2ade

To fix this, we could call `refetch` again to get the latest data whenever a screen is visible again. However, calling `refetch` usually triggers another API call.

So how could we get a screen/component to re-render/get the latest data from cache?

- A new method returned by `useQuery`?  
  - A `forceUpdate` method?
- A new `refetch` option?
  - A `preferCache` option?
    - This option could check if there is something on cache, and if there is, return that instead.

What are your thoughts @TkDodo?

Thank you once again for all you support! 🙇

**Reply by ferrannp** [reply]
I would vote for options two.

```js
refetch({preferCache: true})
```

^ in this case, if there is data in the cache, force a re-render (to get the data) but do not perform a request. If we pass `preferCache: true` but for some reason there is no cache, perform the request.

**Reply by TkDodo** [reply]
if you only need to trigger a re-render, please do that with react state.

**Reply by pedrosimao** [reply]
Seems like a simple enough solution indeed...
Great PR btw!
